doc/developer/guide-changes.md

```diff
@@ -0,0 +1,790 @@
+without requiring a reader to extract this from your code directly. It serves
+as a contract between your implementation and the users of your code. The value
+of clear documentation lies in the extent to which it can reduce the time others
+must spend inspecting your code to determine its appropriateness. Documentation
+also provides for continuity as others may need to work with and take ownership
+of code that you have written.
+
+A PR without appropriate code documentation may reasonably be rejected. Imagine
+that the reviewer would first like to read a precis of your new code, and only
+the read the code itself. This approximates the experience of most people new
+to your new code.
+
+User documentation is also important, but may be outside the tradecraft that
+you are trained for. If you are creating a new user-facing feature, please
+ensure that appropriate documentation is produced, either as part of the pull
+request if that is possible, or scheduled appropriately as the feature is
+released.
+
+### Testing
+
+With perfect testing, introducing a bug in any line of any function in the
+codebase would cause at least one test to fail. Meeting this standard would
+unacceptably impair velocity, but it's a useful framing for what ideal test
+coverage means.
+
+The weaker property we strive for is that every change in behavior should
+require modifying or adding at _least_ one test. If a PR changes, say, the
+behavior of a SQL function and there are no changes to any test files, then
+there are certainly some missing tests.
+
+Whenever possible, tests should validate observable behavior, not implementation
+details. Over-testing of trivial internal functions creates its own maintenance
+burden, as tests frequently have to be changed because of innocuous refactors,
+causing people to be less likely to attempt refactors, or worse, less likely to
+take test failures seriously.
+
+If a piece of code is large and complex enough, it might warrant tests to
+validate its behavior, despite not being externally observable. Even in this
+case, tests should validate the behavior at a clean API boundary of this code,
+rather than its internal details. Always use your best judgment.
+
+Our testing philosophy is covered in much greater detail in
+[Developer guide: testing](guide-testing.md).
+
+## Deviations from Google
+
+As mentioned at the start of this guide, much of the above advice derives from
+Google's [Code Review Developer Guide][google-guide]. There are no notable
+conflicts between Google's guide and ours.
+
+The only major deviation is that we permit PRs to contain multiple commits
+when using the ["semantic commits" strategy](#commit-strategy), whereas Google's
+PR equivalent, the CL ("change list"), always contains exactly one commit.
+As far as we know, Google operates this way because a technical limitation
+became a philosophy, not because restricting a CL to one commit is actually
+preferable.
```

Comment:
nit: if this is referring to gerrit's "change list is a list of changes", this philosophy comes from the fact that gerrit makes reviewing individual commits trivial and correct while github makes it impossible. If you want to review a branch merge in gerrit, the *only* way to do it (from my experience, possibly we had a customized install) is by reviewing each commit ("change") individually.

So philosophically I'd say that gerrit is just the semantic-commits philsophy, enabled at a technical level.

Reply:
Hmm, maybe I have no idea what I'm talking about then. I'm just going to strike the last sentence in this paragraph in the interest of correctness.